src/java/org/apache/cassandra/db/tries/MemtableTrie.java

```diff
@@ -0,0 +1,1039 @@
+                    // we don't want to invalidate the value in other cores' caches unnecessarily).
+                    root = updatedPreContentNode;
+                }
+                return false;
+            }
+            if (updatedPreContentNode != existingPreContentNode)
+                attachChild(transition(), updatedPreContentNode);
+            return true;
+        }
+    }
+
+    /**
+     * Somewhat similar to {@link MergeResolver}, this encapsulates logic to be applied whenever new content is being
+     * upserted into a {@link MemtableTrie}. Unlike {@link MergeResolver}, {@link UpsertTransformer} will be applied no
+     * matter if there's pre-existing content for that trie key/path or not.
+     *
+     * @param <T> The content type for this {@link MemtableTrie}.
+     * @param <U> The type of the new content being applied to this {@link MemtableTrie}.
+     */
+    public interface UpsertTransformer<T, U>
+    {
+        /**
+         * Called when there's content in the updating trie.
+         *
+         * @param existing Existing content for this key, or null if there isn't any.
+         * @param update   The update, always non-null.
+         * @return The combined value to use.
+         */
+        T apply(T existing, U update);
+    }
+
+    /**
+     * Modify this trie to apply the mutation given in the form of a trie. Any content in the mutation will be resolved
+     * with the given function before being placed in this trie (even if there's no pre-existing content in this trie).
+     * @param mutation the mutation to be applied, given in the form of a trie. Note that its content can be of type
+     * different than the element type for this memtable trie.
+     * @param transformer a function applied to the potentially pre-existing value for the given key, and the new
+     * value. Applied even if there's no pre-existing value in the memtable trie.
+     */
+    public <U> void apply(Trie<U> mutation, final UpsertTransformer<T, U> transformer) throws SpaceExhaustedException
+    {
+        Cursor<U> mutationCursor = mutation.cursor();
+        assert mutationCursor.depth() == 0;
+        ApplyState state = applyState;
+        state.reset();
+        state.descend(-1, mutationCursor.content(), transformer);
+        assert state.currentDepth == 0;
+
+        while (true)
+        {
+            int depth = mutationCursor.advance();
+            while (state.currentDepth >= depth)
+            {
+                // There are no more children. Ascend to the parent state to continue walk.
+                if (!state.attachAndMoveToParentState())
+                {
+                    assert depth == -1;
+                    return;
+                }
+            }
+
+            // We have a transition, get child to descend into
+            state.descend(mutationCursor.incomingTransition(), mutationCursor.content(), transformer);
+            assert state.currentDepth == depth;
```

Comment:
nit: In general, would urge brief messages for asserts that survive and may be hit in production code.